### PR TITLE
Refactoring of data loaders

### PR DESCRIPTION
**Pull request by albertogaspar**
- move from callables loaders to iterables
- make the loader more modular
- reduce duplicated code
- removed from_iterable method from loaders

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [X] Other (please describe):

## What is the current behavior?

The `DataLoader`, `InputsLoader`, `TargetsLoader` are hard to extend to new uses cases where the input is not just a batch of `jax.Array` but something else (e.g., when using the `transfromers` library the input data is a `Dict[str, jax.Array]`)

Issue Number: N/A

## What is the new behavior?

- Creation of a general Abstract Base Class for all kinds of loaders
- Make the current `DataLoader`, `InputsLoader`, `TargetsLoader` a concrete implementation of the abstract classes that works with arrays only.

I beleive the proposed implementation is more flexible and allow for easier extendability.

### CODE REVIEW

File: fortuna/data/loader/base.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+            self: T,
+            transform: Callable[[InputData, Status], Tuple[InputData, Status]],
+            status: Optional[Status] = None,
+    ) -> T:
+        """
+        From an existing loader of inputs, create a loader with transformed inputs.
+
+        Parameters
+        ----------
+        transform : Callable[[Array, Status], Tuple[Array, Status]]
```
Comment (gianlucadetommaso) [inline fortuna/data/loader/base.py:253]: InputData instead of Array

File: fortuna/data/loader/base.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+        ],
+    ) -> T:
+        """
+        Transform a callable iterable into a concrete instance of a subclass of 
+        :class:`~fortuna.data.loader.BaseDataLoader`.
+
+        Parameters
+        ----------
+        fun: Callable[[], Iterable[Batch]]
+            A callable iterable of tuples of input and target arrays.
```
Comment (gianlucadetommaso) [inline fortuna/data/loader/base.py:81]: Non necessarily arrays any more?

**Review by gianlucadetommaso — COMMENTED** [review]
All good! Can you just make sure in the docstrings that, for methods that have been generalized to any type, we don't still write "array"?

File: fortuna/data/loader/base.py (hunk 2 above)
Comment (albertogaspar) [reply fortuna/data/loader/base.py:81]: That's right. We should allow fortuna to work with other types of input data (e.g., when using the `transfromers` library the input data is a `Dict[str, jax.Array]`)

**Pull request merged**